// cayenne-lpp/src/datatypes.js
module.exports.Ranges = Object.freeze({
  DIGITAL_INPUT: [0, 255],
  DIGITAL_OUTPUT: [0, 255],
  ANALOG_INPUT: [-327.68, 327.67],
  ANALOG_OUTPUT: [-327.68, 327.67],
  LUMINOSITY: [0, 65535],
  PRESENCE: [0, 1],
  TEMPERATURE: [-3276.8, 3276.7],
  RELATIVE_HUMIDITY: [0, 100],
  ACCELEROMETER: [-32.768, 32.767],
  MAGNETOMETER: [-32.768, 32.767],
  BAROMETRIC_PRESSURE: [0, 6553.5],
  GYROMETER: [-327.68, 327.67],
  GPS_LATITUDE: [-90.0, 90.0],
  GPS_LONGITUDE: [-180.0, 180.0],
  GPS_ALTITUDE: [-3276.8, 3276.7],
});

module.exports.Scales = Object.freeze({
  DIGITAL_INPUT: 1,
  DIGITAL_OUTPUT: 1,
  ANALOG_INPUT: 100,
  ANALOG_OUTPUT: 100,
  LUMINOSITY: 1,
  PRESENCE: 1,
  TEMPERATURE: 10,
  RELATIVE_HUMIDITY: 2,
  ACCELEROMETER: 1000,
  MAGNETOMETER: 1000,
  BAROMETRIC_PRESSURE: 10,
  GYROMETER: 100,
  GPS_LATITUDE: 10000,
  GPS_LONGITUDE: 10000,
  GPS_ALTITUDE: 100,
});

module.exports.Sizes = Object.freeze({
  DIGITAL_INPUT: 1,
  DIGITAL_OUTPUT: 1,
  ANALOG_INPUT: 2,
  ANALOG_OUTPUT: 2,
  LUMINOSITY: 2,
  PRESENCE: 1,
  TEMPERATURE: 2,
  RELATIVE_HUMIDITY: 1,
  ACCELEROMETER: 6,
  MAGNETOMETER: 6,
  BAROMETRIC_PRESSURE: 2,
  GYROMETER: 6,
  GPS: 9,
  CHANNEL_TYPE: 2,
  UINT8: 1,
  UINT16: 2,
  INT16: 2,
  THREE_AXIS: 6,
});

module.exports.Types = Object.freeze({
  DIGITAL_INPUT: 0,
  DIGITAL_OUTPUT: 1,
  ANALOG_INPUT: 2,
  ANALOG_OUTPUT: 3,
  LUMINOSITY: 101,
  PRESENCE: 102,
  TEMPERATURE: 103,
  RELATIVE_HUMIDITY: 104,
  ACCELEROMETER: 113,
  MAGNETOMETER: 114,
  BAROMETRIC_PRESSURE: 115,
  GYROMETER: 134,
  GPS: 136,
});
